source/common/upstream/edf_scheduler.h

```diff
@@ -0,0 +1,101 @@
+#include <cstdint>
+#include <queue>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Upstream {
+
+// It's not sufficient to use trace level logging, since it becomes far too noisy for a number of
+// tests, so we can kill trace debug here.
+#define EDF_DEBUG 0
+
+#if EDF_DEBUG
+#define EDF_TRACE(fmt...) ENVOY_LOG_MISC(trace, fmt)
+#else
+#define EDF_TRACE(fmt...)
+#endif
+
+// Earliest Deadline First (EDF) scheduler
+// (https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) used for weighted round robin.
+// Each pick from the schedule has the earliest deadline entry selected. Entries have deadlines set
+// at current time + 1 / weight, providing weighted round robin behavior with floating point
```

Comment:
To be perfectly honest I'm really only reviewing this all for code sanity. I don't feel comfortable really reviewing this algorithm. Assuming you got this method from a reliable source (which I think you did), I'm fine with it. But up to you if you want to try to find someone more capable to review the algorithm.

Reply:
@alyssawilk do you think we should stick with EDF, or do WRR the spacey way (i.e. build an explicit schedule such as [1,1,2,1,1,3,1,1,2,3] and cycles through it)? If we do it the spacey way, we are not going to be able to do arbitrary precision weighting or large weight ranges. I think the explicit schedule is much simpler to understand and we can switch to it fairly easily, but I would rather not throw away the EDF scheduler if we will need larger ranges or float precision later in WRR.

Reply:
I think EDF is well tested in house, and as we're the main customer of the complicated LB code I think we might as well do the version we know works well and scales well.  I think it's totally fair for Matt to say he's not going to take the time to correctness check - we're going to be the ones debugging it if there's bugs or perf problems anyway so on our heads be it if it's not correct :-)

Reply:
I'm no expert, but I stared at it long enough to understand how it works and I'm satisfied that it works as indicated.